// @docker/desktop-theme/src/colors.ts
/* defined here: https://www.figma.com/file/rien0oukkvklddrnv8akct/all-colors_dds-(copy) */

export const colors = {
  red10: '#F8A09A',
  red20: '#F3847E',
  red30: '#EE635F',
  red40: '#E9524F',
  red50: '#E44141',
  red60: '#DA3333',
  red70: '#C82828',
  red80: '#B21C1C',
  red90: '#9E1011',
  red100: '#8C0607' /* a.k.a. docker red */,

  magenta10: '#FBD5FA',
  magenta20: '#F5B8F3',
  magenta30: '#EF94EB',
  magenta40: '#E96FE4',
  magenta50: '#E34ADC',
  magenta60: '#DD26D5',
  magenta70: '#CC18C4',
  magenta80: '#A611A0',
  magenta90: '#7B0A76',
  magenta100: '#530450',

  purple10: '#D4C1F8',
  purple20: '#BCA5EA',
  purple30: '#AA92DC',
  purple40: '#987ECE',
  purple50: '#866ABF',
  purple60: '#7457B1',
  purple70: '#6243A3',
  purple80: '#4E2D8F',
  purple90: '#3C1980',
  purple100: '#270769',

  marine10: '#9FD1F9',
  marine20: '#6FB6F6',
  marine30: '#4BA0F3',
  marine40: '#268AEF',
  marine50: '#007BFF',
  marine60: '#0162CC',
  marine70: '#0151AD',
  marine80: '#003F8C',
  marine90: '#002C66',
  marine100: '#0B214A',

  grey10: '#D7DADE',
  grey20: '#B9C2C9',
  grey30: '#A5B1BA',
  grey40: '#90A0AC',
  grey50: '#768491',
  grey60: '#5A6774',
  grey70: '#404D59',
  grey80: '#26323F',
  grey90: '#17222F',
  grey100: '#161F2A',

  blue10: '#C4F0FF',
  blue20: '#98E0FF',
  blue30: '#71CFFF',
  blue40: '#51BCFD',
  blue50: '#37AAF7',
  blue60: '#2496ED' /* a.k.a. docker blue */,
  blue70: '#1282D7',
  blue80: '#076CAD',
  blue90: '#025391',
  blue100: '#004375',

  green10: '#C7EFE6',
  green20: '#AAE8D8',
  green30: '#8DE0CA',
  green40: '#70D8BD',
  green50: '#54D1B0' /* a.k.a. docker green */,
  green60: '#43B899',
  green70: '#33977E',
  green80: '#237663',
  green90: '#135446',
  green100: '#094539',

  orange10: '#FEF5C3',
  orange20: '#F9E3AE',
  orange30: '#F5D19A',
  orange40: '#F0BF85',
  orange50: '#ECAD71',
  orange60: '#E79A5C',
  orange70: '#E28847',
  orange80: '#DB7432',
  orange90: '#D4611D',
  orange100: '#CC4E09',

  white0: '#FFFFFF',
  white5: '#FAFAFB',
  white10: '#F7F7F8',
  white15: '#F2F3F5',
  white25: '#E9EBEE',

  black0: '#0F161E',
  black5: '#121A24',
  black15: '#151C26',
};
