// @docker/desktop-theme/src/docker-color-palettes.ts
/*

The below content was sourced from Figma:
https://www.figma.com/file/YrCXp3zztpwtcSD9XVge7Z/%E2%9C%A8Foundations?node-id=0%3A1

PLEASE DON'T use logic like "use red 200 in dark theme and red 600 in light theme"

The idea behind these themes is that they're supposed to be drop-in replacements
for each other. Just pick a hue and a number and use them in both dark and light
modes, and trust the palette to give you something that looks good.

If the palette betrays you and gives you something that looks bad, it might be
because the UI around your component is not using the palette. In this case,
check with designers about how you can bring that surrounding UI up to date.

*/

export interface DockerPalette {
  readonly amber: DockerPaletteHue;
  readonly blue: DockerPaletteHue;
  readonly green: DockerPaletteHue;
  readonly grey: DockerPaletteHue;
  readonly red: DockerPaletteHue;
  readonly violet: DockerPaletteHue;
  readonly background: string;
  readonly text: string;
}

export interface DockerPaletteHue {
  /** when we only had one palette, this was called "tint" */
  readonly '100': string;

  /** when we only had one palette, this was called "lightest" */
  readonly '200': string;

  /** when we only had one palette, this was called "lighter" */
  readonly '300': string;

  /** when we only had one palette, this was called "light" */
  readonly '400': string;

  /** The primary instance of this hue. When we only had one palette, this was called "neutral" */
  readonly '500': string;

  /** when we only had one palette, this was called "dark" */
  readonly '600': string;

  /** when we only had one palette, this was called "darker" */
  readonly '700': string;

  /** when we only had one palette, this was called "darkest" */
  readonly '800': string;
}

export const dockerLightPalette: DockerPalette = {
  // contrast colors:
  // 100-400: black
  // 500-800: white
  amber: {
    '100': '#FFF6E2',
    '200': '#FCE1A9',
    '300': '#FBB552',
    '400': '#DD7805',
    '500': '#B85504',
    '600': '#AA4409',
    '700': '#893607',
    '800': '#421A02',
  },
  blue: {
    '100': '#E8F4FD',
    '200': '#C0E0FA',
    '300': '#8BC7F5',
    '400': '#1C90ED',
    '500': '#116ED0',
    '600': '#0055BD',
    '700': '#1144A6',
    '800': '#00084D',
  },
  green: {
    '100': '#F3FAF9',
    '200': '#C6EAE1',
    '300': '#88D5C0',
    '400': '#3BA08D',
    '500': '#2E7F74',
    '600': '#1E6C5F',
    '700': '#185A51',
    '800': '#0C2C28',
  },
  grey: {
    '100': '#EAEAED',
    '200': '#E1E2E6',
    '300': '#C4C8D1',
    '400': '#8993A5',
    '500': '#677285',
    '600': '#505968',
    '700': '#393F49',
    '800': '#17191E',
  },
  red: {
    '100': '#FEEFEF',
    '200': '#F6CFD0',
    '300': '#EEA3A5',
    '400': '#E25D68',
    '500': '#D52536',
    '600': '#B72132',
    '700': '#8B1924',
    '800': '#350A10',
  },
  violet: {
    '100': '#F7ECFF',
    '200': '#E9D4FF',
    '300': '#C9A6FF',
    '400': '#9860FF',
    '500': '#7D2EFF',
    '600': '#6D00EB',
    '700': '#5700BB',
    '800': '#220041',
  },
  // this isn't per the design system, it's just copied from
  // existing DD style until we can get the right color
  text: '#27343B',
  background: '#F9F9FA',
};

export const dockerDarkPalette: DockerPalette = {
  // contrast colors:
  // 100-400: white
  // 500-800: black
  amber: {
    '100': '#672805',
    '200': '#7B3606',
    '300': '#944307',
    '400': '#AF560A',
    '500': '#CD6A0A',
    '600': '#ED8D25',
    '700': '#F6A650',
    '800': '#F8B974',
  },
  blue: {
    '100': '#0B1B46',
    '200': '#06326D',
    '300': '#084391',
    '400': '#1557B8',
    '500': '#116ED0',
    '600': '#3391EE',
    '700': '#55A4F1',
    '800': '#7CB9F4',
  },
  green: {
    '100': '#0D342E',
    '200': '#11423B',
    '300': '#145348',
    '400': '#1A655A',
    '500': '#228375',
    '600': '#2AA391',
    '700': '#3CC1AD',
    '800': '#7ACCC3',
  },
  grey: {
    '100': '#141B1F',
    '200': '#27343E',
    '300': '#364754',
    '400': '#465C6E',
    '500': '#7794AB',
    '600': '#94ABBC',
    '700': '#ADBECB',
    '800': '#C4D0DA',
  },
  red: {
    '100': '#58111B',
    '200': '#741624',
    '300': '#951C2F',
    '400': '#BC233C',
    '500': '#C32438',
    '600': '#DD4659',
    '700': '#E36676',
    '800': '#EA8E9A',
  },
  violet: {
    '100': '#341458',
    '200': '#491D7B',
    '300': '#5F25A0',
    '400': '#8032D6',
    '500': '#8A53EC',
    '600': '#A371FC',
    '700': '#B38BFC',
    '800': '#C5A6FD',
  },
  text: '#FFFFFF',
  background: '#1C262D',
};
