// @docker/extension-api-client-ipc/dist/createDesktopUINode.js
export function createDesktopUINode(ipcRenderer) {
    return {
        toast: {
            error(msg) {
                ipcRenderer.invoke(`display-toast-error`, msg);
            },
            success(msg) {
                ipcRenderer.invoke(`display-toast-success`, msg);
            },
            warning(msg) {
                ipcRenderer.invoke(`display-toast-warning`, msg);
            },
        },
        dialog: {
            showOpenDialog(dialogProperties) {
                return ipcRenderer.invoke(`open-dialog`, dialogProperties);
            },
        },
        navigate: {
            async viewContainers() {
                await ipcRenderer.invoke(`navigate-to-containers`);
            },
            async viewContainer(id) {
                await ipcRenderer.invoke(`navigate-to-container`, { id });
            },
            async viewContainerLogs(id) {
                await ipcRenderer.invoke(`navigate-to-container-logs`, { id });
            },
            async viewContainerInspect(id) {
                await ipcRenderer.invoke(`navigate-to-container-inspect`, { id });
            },
            async viewContainerStats(id) {
                await ipcRenderer.invoke(`navigate-to-container-stats`, { id });
            },
            async viewImages() {
                await ipcRenderer.invoke(`navigate-to-images`);
            },
            async viewImage(id, tag) {
                await ipcRenderer.invoke(`navigate-to-image`, { id, tag });
            },
            async viewVolumes() {
                await ipcRenderer.invoke(`navigate-to-volumes`);
            },
            async viewVolume(volume) {
                await ipcRenderer.invoke(`navigate-to-volume`, volume);
            },
            async viewDevEnvironments() {
                await ipcRenderer.invoke(`navigate-to-devenvs`);
            },
        },
    };
}
