// @mui/x-data-grid-pro/legacy/hooks/features/columnPinning/useGridColumnPinning.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _extends from "@babel/runtime/helpers/esm/extends";
import * as React from 'react';
import MuiDivider from '@mui/material/Divider';
import { useGridSelector, gridVisibleColumnDefinitionsSelector, gridColumnsTotalWidthSelector, gridColumnPositionsSelector, gridVisibleColumnFieldsSelector, gridClasses, GridEvents, useGridApiMethod, useGridApiEventHandler } from '@mui/x-data-grid';
import { useGridRegisterPreProcessor } from '@mui/x-data-grid/internals';
import { GridColumnPinningMenuItems } from '../../../components/GridColumnPinningMenuItems';
import { GridPinnedPosition } from './gridColumnPinningInterface';
import { gridPinnedColumnsSelector } from './gridColumnPinningSelector';
import { filterColumns } from '../../../components/DataGridProVirtualScroller';
import { jsx as _jsx } from "react/jsx-runtime";

var Divider = function Divider() {
  return /*#__PURE__*/_jsx(MuiDivider, {
    onClick: function onClick(event) {
      return event.stopPropagation();
    }
  });
};

export var columnPinningStateInitializer = function columnPinningStateInitializer(state, props) {
  var _props$initialState;

  var model;

  if (props.disableColumnPinning) {
    model = {};
  } else if (props.pinnedColumns) {
    model = props.pinnedColumns;
  } else if ((_props$initialState = props.initialState) != null && _props$initialState.pinnedColumns) {
    var _props$initialState2;

    model = (_props$initialState2 = props.initialState) == null ? void 0 : _props$initialState2.pinnedColumns;
  } else {
    model = {};
  }

  return _extends({}, state, {
    pinnedColumns: model
  });
};

var mergeStateWithPinnedColumns = function mergeStateWithPinnedColumns(pinnedColumns) {
  return function (state) {
    return _extends({}, state, {
      pinnedColumns: pinnedColumns
    });
  };
};

export var useGridColumnPinning = function useGridColumnPinning(apiRef, props) {
  var pinnedColumns = useGridSelector(apiRef, gridPinnedColumnsSelector); // Each visible row (not to be confused with a filter result) is composed of a central .MuiDataGrid-row element
  // and up to two additional .MuiDataGrid-row's, one for the columns pinned to the left and another
  // for those on the right side. When hovering any of these elements, the :hover styles are applied only to
  // the row element that was actually hovered, not its additional siblings. To make it look like a contiguous row,
  // this method adds/removes the .Mui-hovered class to all of the row elements inside one visible row.

  var updateHoveredClassOnSiblingRows = React.useCallback(function (event) {
    var _pinnedColumns$left$l, _pinnedColumns$left, _pinnedColumns$right$, _pinnedColumns$right;

    if (props.disableColumnPinning) {
      return;
    }

    if (!Array.isArray(pinnedColumns.left) && !Array.isArray(pinnedColumns.right)) {
      return;
    }

    var nbLeftPinnedColumns = (_pinnedColumns$left$l = (_pinnedColumns$left = pinnedColumns.left) == null ? void 0 : _pinnedColumns$left.length) != null ? _pinnedColumns$left$l : 0;
    var nbRightPinnedColumns = (_pinnedColumns$right$ = (_pinnedColumns$right = pinnedColumns.right) == null ? void 0 : _pinnedColumns$right.length) != null ? _pinnedColumns$right$ : 0;

    if (nbLeftPinnedColumns + nbRightPinnedColumns === 0) {
      return;
    }

    var index = event.currentTarget.dataset.rowindex;
    var rowElements = apiRef.current.windowRef.current.querySelectorAll(".".concat(gridClasses.row, "[data-rowindex=\"").concat(index, "\"]"));
    rowElements.forEach(function (row) {
      // Ignore rows from other grid inside the hovered row
      if (row.closest(".".concat(gridClasses.virtualScroller)) === apiRef.current.windowRef.current) {
        if (event.type === 'mouseenter') {
          row.classList.add('Mui-hovered');
        } else {
          row.classList.remove('Mui-hovered');
        }
      }
    });
  }, [apiRef, pinnedColumns.left, pinnedColumns.right, props.disableColumnPinning]);
  var handleMouseEnter = React.useCallback(function (params, event) {
    updateHoveredClassOnSiblingRows(event);
  }, [updateHoveredClassOnSiblingRows]);
  var handleMouseLeave = React.useCallback(function (params, event) {
    updateHoveredClassOnSiblingRows(event);
  }, [updateHoveredClassOnSiblingRows]);
  useGridApiEventHandler(apiRef, GridEvents.rowMouseEnter, handleMouseEnter);
  useGridApiEventHandler(apiRef, GridEvents.rowMouseLeave, handleMouseLeave);
  /**
   * PRE-PROCESSING
   */

  var calculateScrollLeft = React.useCallback(function (initialValue, params) {
    if (props.disableColumnPinning) {
      return initialValue;
    }

    var visibleColumnFields = gridVisibleColumnFieldsSelector(apiRef);

    var _filterColumns = filterColumns(pinnedColumns, visibleColumnFields),
        _filterColumns2 = _slicedToArray(_filterColumns, 2),
        leftPinnedColumns = _filterColumns2[0],
        rightPinnedColumns = _filterColumns2[1];

    if (!params.colIndex || leftPinnedColumns.length === 0 && rightPinnedColumns.length === 0) {
      return initialValue;
    }

    var visibleColumns = gridVisibleColumnDefinitionsSelector(apiRef);
    var columnsTotalWidth = gridColumnsTotalWidthSelector(apiRef);
    var columnPositions = gridColumnPositionsSelector(apiRef);
    var clientWidth = apiRef.current.windowRef.current.clientWidth;
    var scrollLeft = apiRef.current.windowRef.current.scrollLeft;
    var offsetWidth = visibleColumns[params.colIndex].computedWidth;
    var offsetLeft = columnPositions[params.colIndex];
    var leftPinnedColumnsWidth = columnPositions[leftPinnedColumns.length];
    var rightPinnedColumnsWidth = columnsTotalWidth - columnPositions[columnPositions.length - rightPinnedColumns.length];
    var elementBottom = offsetLeft + offsetWidth;

    if (elementBottom - (clientWidth - rightPinnedColumnsWidth) > scrollLeft) {
      var left = elementBottom - (clientWidth - rightPinnedColumnsWidth);
      return _extends({}, initialValue, {
        left: left
      });
    }

    if (offsetLeft < scrollLeft + leftPinnedColumnsWidth) {
      var _left = offsetLeft - leftPinnedColumnsWidth;

      return _extends({}, initialValue, {
        left: _left
      });
    }

    return initialValue;
  }, [apiRef, pinnedColumns, props.disableColumnPinning]);
  var addColumnMenuButtons = React.useCallback(function (initialValue, column) {
    if (props.disableColumnPinning) {
      return initialValue;
    }

    if (column.pinnable === false) {
      return initialValue;
    }

    return [].concat(_toConsumableArray(initialValue), [/*#__PURE__*/_jsx(Divider, {}), /*#__PURE__*/_jsx(GridColumnPinningMenuItems, {})]);
  }, [props.disableColumnPinning]);
  var checkIfCanBeReordered = React.useCallback(function (initialValue, _ref) {
    var targetIndex = _ref.targetIndex;
    var visibleColumnFields = gridVisibleColumnFieldsSelector(apiRef);

    var _filterColumns3 = filterColumns(pinnedColumns, visibleColumnFields),
        _filterColumns4 = _slicedToArray(_filterColumns3, 2),
        leftPinnedColumns = _filterColumns4[0],
        rightPinnedColumns = _filterColumns4[1];

    if (leftPinnedColumns.length === 0 && rightPinnedColumns.length === 0) {
      return initialValue;
    }

    if (leftPinnedColumns.length > 0 && targetIndex < leftPinnedColumns.length) {
      return false;
    }

    if (rightPinnedColumns.length > 0) {
      var visibleColumns = gridVisibleColumnDefinitionsSelector(apiRef);
      var firstRightPinnedColumnIndex = visibleColumns.length - rightPinnedColumns.length;
      return targetIndex >= firstRightPinnedColumnIndex ? false : initialValue;
    }

    return initialValue;
  }, [apiRef, pinnedColumns]);
  var stateExportPreProcessing = React.useCallback(function (prevState) {
    var pinnedColumnsToExport = gridPinnedColumnsSelector(apiRef.current.state);

    if ((!pinnedColumnsToExport.left || pinnedColumnsToExport.left.length === 0) && (!pinnedColumnsToExport.right || pinnedColumnsToExport.right.length === 0)) {
      return prevState;
    }

    return _extends({}, prevState, {
      pinnedColumns: pinnedColumnsToExport
    });
  }, [apiRef]);
  var stateRestorePreProcessing = React.useCallback(function (params, context) {
    var newPinnedColumns = context.stateToRestore.pinnedColumns;

    if (newPinnedColumns != null) {
      apiRef.current.setState(mergeStateWithPinnedColumns(newPinnedColumns));
    }

    return params;
  }, [apiRef]);
  useGridRegisterPreProcessor(apiRef, 'scrollToIndexes', calculateScrollLeft);
  useGridRegisterPreProcessor(apiRef, 'columnMenu', addColumnMenuButtons);
  useGridRegisterPreProcessor(apiRef, 'canBeReordered', checkIfCanBeReordered);
  useGridRegisterPreProcessor(apiRef, 'exportState', stateExportPreProcessing);
  useGridRegisterPreProcessor(apiRef, 'restoreState', stateRestorePreProcessing);
  apiRef.current.unstable_updateControlState({
    stateId: 'pinnedColumns',
    propModel: props.pinnedColumns,
    propOnChange: props.onPinnedColumnsChange,
    stateSelector: gridPinnedColumnsSelector,
    changeEvent: GridEvents.pinnedColumnsChange
  });
  var checkIfEnabled = React.useCallback(function (methodName) {
    if (props.disableColumnPinning) {
      throw new Error("MUI: You cannot call `apiRef.current.".concat(methodName, "` when `disableColumnPinning` is true."));
    }
  }, [props.disableColumnPinning]);
  var pinColumn = React.useCallback(function (field, side) {
    checkIfEnabled('pinColumn');

    if (apiRef.current.isColumnPinned(field) === side) {
      return;
    }

    apiRef.current.setState(function (state) {
      var _extends2;

      var otherSide = side === GridPinnedPosition.right ? GridPinnedPosition.left : GridPinnedPosition.right;

      var newPinnedColumns = _extends({}, state.pinnedColumns, (_extends2 = {}, _defineProperty(_extends2, side, [].concat(_toConsumableArray(state.pinnedColumns[side] || []), [field])), _defineProperty(_extends2, otherSide, (state.pinnedColumns[otherSide] || []).filter(function (column) {
        return column !== field;
      })), _extends2));

      return _extends({}, state, {
        pinnedColumns: newPinnedColumns
      });
    });
    apiRef.current.forceUpdate();
  }, [apiRef, checkIfEnabled]);
  var unpinColumn = React.useCallback(function (field) {
    checkIfEnabled('unpinColumn');
    apiRef.current.setState(function (state) {
      var newPinnedColumns = _extends({}, state.pinnedColumns, {
        left: (state.pinnedColumns.left || []).filter(function (column) {
          return column !== field;
        }),
        right: (state.pinnedColumns.right || []).filter(function (column) {
          return column !== field;
        })
      });

      return _extends({}, state, {
        pinnedColumns: newPinnedColumns
      });
    });
    apiRef.current.forceUpdate();
  }, [apiRef, checkIfEnabled]);
  var getPinnedColumns = React.useCallback(function () {
    checkIfEnabled('getPinnedColumns');
    return gridPinnedColumnsSelector(apiRef.current.state);
  }, [apiRef, checkIfEnabled]);
  var setPinnedColumns = React.useCallback(function (newPinnedColumns) {
    checkIfEnabled('setPinnedColumns');
    apiRef.current.setState(mergeStateWithPinnedColumns(newPinnedColumns));
    apiRef.current.forceUpdate();
  }, [apiRef, checkIfEnabled]);
  var isColumnPinned = React.useCallback(function (field) {
    checkIfEnabled('isColumnPinned');
    var leftPinnedColumns = pinnedColumns.left || [];

    if (leftPinnedColumns.includes(field)) {
      return GridPinnedPosition.left;
    }

    var rightPinnedColumns = pinnedColumns.right || [];

    if (rightPinnedColumns.includes(field)) {
      return GridPinnedPosition.right;
    }

    return false;
  }, [pinnedColumns.left, pinnedColumns.right, checkIfEnabled]);
  var columnPinningApi = {
    pinColumn: pinColumn,
    unpinColumn: unpinColumn,
    getPinnedColumns: getPinnedColumns,
    setPinnedColumns: setPinnedColumns,
    isColumnPinned: isColumnPinned
  };
  useGridApiMethod(apiRef, columnPinningApi, 'columnPinningApi');
  React.useEffect(function () {
    if (props.pinnedColumns) {
      apiRef.current.setPinnedColumns(props.pinnedColumns);
    }
  }, [apiRef, props.pinnedColumns]);
};
