// @mui/x-data-grid-pro/legacy/components/DataGridProColumnHeaders.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import _extends from "@babel/runtime/helpers/esm/extends";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
var _excluded = ["style", "className", "innerRef"];
import * as React from 'react';
import { unstable_composeClasses as composeClasses } from '@mui/material';
import { styled, alpha } from '@mui/material/styles';
import { getDataGridUtilityClass, gridClasses, useGridSelector, useGridApiEventHandler, gridVisibleColumnFieldsSelector, GridEvents, GridColumnHeaderSeparatorSides } from '@mui/x-data-grid';
import { GridColumnHeaders, GridColumnHeadersInner, useGridColumnHeaders } from '@mui/x-data-grid/internals';
import { useGridRootProps } from '../hooks/utils/useGridRootProps';
import { useGridApiContext } from '../hooks/utils/useGridApiContext';
import { gridPinnedColumnsSelector, GridPinnedPosition } from '../hooks/features/columnPinning';
import { filterColumns } from './DataGridProVirtualScroller';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";

var useUtilityClasses = function useUtilityClasses(ownerState) {
  var leftPinnedColumns = ownerState.leftPinnedColumns,
      rightPinnedColumns = ownerState.rightPinnedColumns,
      classes = ownerState.classes;
  var slots = {
    leftPinnedColumns: ['pinnedColumnHeaders', leftPinnedColumns && leftPinnedColumns.length > 0 && "pinnedColumnHeaders--left"],
    rightPinnedColumns: ['pinnedColumnHeaders', rightPinnedColumns && rightPinnedColumns.length > 0 && "pinnedColumnHeaders--right"]
  };
  return composeClasses(slots, getDataGridUtilityClass, classes);
};

// Inspired by https://github.com/material-components/material-components-ios/blob/bca36107405594d5b7b16265a5b0ed698f85a5ee/components/Elevation/src/UIColor%2BMaterialElevation.m#L61
var getOverlayAlpha = function getOverlayAlpha(elevation) {
  var alphaValue;

  if (elevation < 1) {
    alphaValue = 5.11916 * Math.pow(elevation, 2);
  } else {
    alphaValue = 4.5 * Math.log(elevation + 1) + 2;
  }

  return alphaValue / 100;
};

var GridColumnHeadersPinnedColumnHeaders = styled('div', {
  name: 'MuiDataGrid',
  slot: 'PinnedColumnHeaders',
  overridesResolver: function overridesResolver(props, styles) {
    return [_defineProperty({}, "&.".concat(gridClasses['pinnedColumnHeaders--left']), styles['pinnedColumnHeaders--left']), _defineProperty({}, "&.".concat(gridClasses['pinnedColumnHeaders--right']), styles['pinnedColumnHeaders--right']), styles.pinnedColumnHeaders];
  }
})(function (_ref3) {
  var theme = _ref3.theme,
      ownerState = _ref3.ownerState;
  return _extends({
    position: 'absolute',
    overflow: 'hidden',
    height: '100%',
    zIndex: 1,
    display: 'flex',
    boxShadow: theme.shadows[2],
    backgroundColor: theme.palette.background.default
  }, theme.palette.mode === 'dark' && {
    backgroundImage: "linear-gradient(".concat(alpha('#fff', getOverlayAlpha(2)), ", ").concat(alpha('#fff', getOverlayAlpha(2)), ")")
  }, ownerState.side === GridPinnedPosition.left && {
    left: 0
  }, ownerState.side === GridPinnedPosition.right && {
    right: 0
  });
});
export var DataGridProColumnHeaders = /*#__PURE__*/React.forwardRef(function DataGridProColumnHeaders(props, ref) {
  var style = props.style,
      className = props.className,
      innerRef = props.innerRef,
      other = _objectWithoutProperties(props, _excluded);

  var rootProps = useGridRootProps();
  var apiRef = useGridApiContext();
  var visibleColumnFields = useGridSelector(apiRef, gridVisibleColumnFieldsSelector);

  var _React$useState = React.useState(0),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      scrollbarSize = _React$useState2[0],
      setScrollbarSize = _React$useState2[1];

  var handleContentSizeChange = React.useCallback(function () {
    var _apiRef$current$windo;

    if (!((_apiRef$current$windo = apiRef.current.windowRef) != null && _apiRef$current$windo.current)) {
      return;
    } // TODO expose scrollbar size on getRootDimensions


    var newScrollbarSize = apiRef.current.windowRef.current.offsetWidth - apiRef.current.windowRef.current.clientWidth;
    setScrollbarSize(newScrollbarSize);
  }, [apiRef]);
  useGridApiEventHandler(apiRef, GridEvents.virtualScrollerContentSizeChange, handleContentSizeChange);
  var pinnedColumns = useGridSelector(apiRef, gridPinnedColumnsSelector);

  var _filterColumns = filterColumns(pinnedColumns, visibleColumnFields),
      _filterColumns2 = _slicedToArray(_filterColumns, 2),
      leftPinnedColumns = _filterColumns2[0],
      rightPinnedColumns = _filterColumns2[1];

  var _useGridColumnHeaders = useGridColumnHeaders({
    innerRef: innerRef,
    minColumnIndex: leftPinnedColumns.length
  }),
      isDragging = _useGridColumnHeaders.isDragging,
      renderContext = _useGridColumnHeaders.renderContext,
      updateInnerPosition = _useGridColumnHeaders.updateInnerPosition,
      getRootProps = _useGridColumnHeaders.getRootProps,
      getInnerProps = _useGridColumnHeaders.getInnerProps,
      getColumns = _useGridColumnHeaders.getColumns;

  var ownerState = {
    leftPinnedColumns: leftPinnedColumns,
    rightPinnedColumns: rightPinnedColumns,
    classes: rootProps.classes
  };
  var classes = useUtilityClasses(ownerState);
  React.useEffect(function () {
    if (renderContext) {
      updateInnerPosition(renderContext);
    }
  }, [renderContext, updateInnerPosition]);
  var leftRenderContext = renderContext && leftPinnedColumns.length ? _extends({}, renderContext, {
    firstColumnIndex: 0,
    lastColumnIndex: leftPinnedColumns.length
  }) : null;
  var rightRenderContext = renderContext && rightPinnedColumns.length ? _extends({}, renderContext, {
    firstColumnIndex: visibleColumnFields.length - rightPinnedColumns.length,
    lastColumnIndex: visibleColumnFields.length
  }) : null;
  return /*#__PURE__*/_jsxs(GridColumnHeaders, _extends({
    ref: ref,
    className: className
  }, getRootProps(other), {
    children: [leftRenderContext && /*#__PURE__*/_jsx(GridColumnHeadersPinnedColumnHeaders, {
      className: classes.leftPinnedColumns,
      ownerState: {
        side: GridPinnedPosition.left
      },
      children: getColumns({
        renderContext: leftRenderContext,
        minFirstColumn: leftRenderContext.firstColumnIndex,
        maxLastColumn: leftRenderContext.lastColumnIndex
      }, {
        disableReorder: true
      })
    }), rightRenderContext && /*#__PURE__*/_jsx(GridColumnHeadersPinnedColumnHeaders, {
      ownerState: {
        side: GridPinnedPosition.right
      },
      className: classes.rightPinnedColumns,
      style: {
        paddingRight: scrollbarSize
      },
      children: getColumns({
        renderContext: rightRenderContext,
        minFirstColumn: rightRenderContext.firstColumnIndex,
        maxLastColumn: rightRenderContext.lastColumnIndex
      }, {
        disableReorder: true,
        separatorSide: GridColumnHeaderSeparatorSides.Left
      })
    }), /*#__PURE__*/_jsx(GridColumnHeadersInner, _extends({
      isDragging: isDragging
    }, getInnerProps(), {
      children: getColumns({
        renderContext: renderContext,
        minFirstColumn: leftPinnedColumns.length,
        maxLastColumn: visibleColumnFields.length - rightPinnedColumns.length
      })
    }))]
  }));
});
