// @mui/x-license-pro/legacy/cli/license-cli.js
/* eslint-disable no-console */
import * as yargs from 'yargs';
import { generateLicence } from '../generateLicense/generateLicense';
var oneDayInMs = 1000 * 60 * 60 * 24;
export function licenseGenCli() {
  yargs.command({
    command: '$0',
    describe: 'Generates Component.propTypes from TypeScript declarations',
    builder: function builder(command) {
      return command.option('order', {
        default: '',
        alias: 'o',
        describe: 'Order number id.',
        type: 'string'
      }).option('expiry', {
        default: '366',
        describe: 'Number of days from now until expiry date.',
        type: 'string'
      });
    },
    handler: function handler(argv) {
      if (!argv.order) {
        throw new Error('MUI: You forgot to pass an order number. $ > licensegen -o order_123.');
      }

      var licenseDetails = {
        expiryDate: new Date(new Date().getTime() + parseInt(argv.expiry, 10) * oneDayInMs),
        orderNumber: argv.order
      };
      console.log("Generating new license number for order ".concat(licenseDetails.orderNumber, " with expiry date ").concat(licenseDetails.expiryDate.toLocaleDateString()));
      var license = generateLicence(licenseDetails);
      console.log("New license: \n".concat(license));
    }
  }).help().strict(true).version(false).parse();
}
