// @mui/x-license-pro/legacy/utils/licenseInfo.js
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";

/* eslint-disable no-underscore-dangle */
import { ponyfillGlobal } from '@mui/utils'; // Store the license information in a global so it can be shared
// when module duplication occurs. The duplication of the modules can happen
// if using multiple version of MUI X at the same time of the bundler
// decide to duplicate to improve the size of the chunks.

ponyfillGlobal.__MUI_LICENSE_INFO__ = ponyfillGlobal.__MUI_LICENSE_INFO__ || {
  key: undefined,
  releaseInfo: undefined
};
export var LicenseInfo = /*#__PURE__*/function () {
  function LicenseInfo() {
    _classCallCheck(this, LicenseInfo);
  }

  _createClass(LicenseInfo, null, [{
    key: "getKey",
    value: function getKey() {
      return ponyfillGlobal.__MUI_LICENSE_INFO__.key;
    }
  }, {
    key: "getReleaseInfo",
    value: function getReleaseInfo() {
      return ponyfillGlobal.__MUI_LICENSE_INFO__.releaseInfo;
    }
  }, {
    key: "setLicenseKey",
    value: function setLicenseKey(key) {
      ponyfillGlobal.__MUI_LICENSE_INFO__.key = key;
    }
  }, {
    key: "setReleaseInfo",
    value: function setReleaseInfo(encodedReleaseInfo) {
      ponyfillGlobal.__MUI_LICENSE_INFO__.releaseInfo = encodedReleaseInfo;
    }
  }]);

  return LicenseInfo;
}();
