// @mui/x-data-grid/legacy/hooks/features/scroll/useGridScroll.js
import _typeof from "@babel/runtime/helpers/esm/typeof";
import * as React from 'react';
import { useGridLogger } from '../../utils/useGridLogger';
import { gridColumnPositionsSelector, gridVisibleColumnDefinitionsSelector } from '../columns/gridColumnsSelector';
import { gridPageSelector, gridPageSizeSelector } from '../pagination/gridPaginationSelector';
import { gridRowCountSelector } from '../rows/gridRowsSelector';
import { gridRowsMetaSelector } from '../rows/gridRowsMetaSelector';
import { useGridApiMethod } from '../../utils/useGridApiMethod';
import { useGridNativeEventListener } from '../../utils/useGridNativeEventListener'; // Logic copied from https://www.w3.org/TR/wai-aria-practices/examples/listbox/js/listbox.js
// Similar to https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView

function scrollIntoView(dimensions) {
  var clientHeight = dimensions.clientHeight,
      scrollTop = dimensions.scrollTop,
      offsetHeight = dimensions.offsetHeight,
      offsetTop = dimensions.offsetTop;
  var elementBottom = offsetTop + offsetHeight;

  if (elementBottom - clientHeight > scrollTop) {
    return elementBottom - clientHeight;
  }

  if (offsetTop < scrollTop) {
    return offsetTop;
  }

  return undefined;
}
/**
 * @requires useGridPagination (state) - can be after, async only
 * @requires useGridColumns (state) - can be after, async only
 * @requires useGridRows (state) - can be after, async only
 * @requires useGridRowsMeta (state) - can be after, async only
 */


export var useGridScroll = function useGridScroll(apiRef, props) {
  var logger = useGridLogger(apiRef, 'useGridScroll');
  var colRef = apiRef.current.columnHeadersElementRef;
  var windowRef = apiRef.current.windowRef;
  var scrollToIndexes = React.useCallback(function (params) {
    var totalRowCount = gridRowCountSelector(apiRef);
    var visibleColumns = gridVisibleColumnDefinitionsSelector(apiRef);

    if (totalRowCount === 0 || visibleColumns.length === 0) {
      return false;
    }

    logger.debug("Scrolling to cell at row ".concat(params.rowIndex, ", col: ").concat(params.colIndex, " "));
    var scrollCoordinates = {};

    if (params.colIndex != null) {
      var columnPositions = gridColumnPositionsSelector(apiRef);
      scrollCoordinates.left = scrollIntoView({
        clientHeight: windowRef.current.clientWidth,
        scrollTop: windowRef.current.scrollLeft,
        offsetHeight: visibleColumns[params.colIndex].computedWidth,
        offsetTop: columnPositions[params.colIndex]
      });
    }

    if (params.rowIndex != null) {
      var rowsMeta = gridRowsMetaSelector(apiRef.current.state);
      var page = gridPageSelector(apiRef);
      var pageSize = gridPageSizeSelector(apiRef);
      var elementIndex = !props.pagination ? params.rowIndex : params.rowIndex - page * pageSize;
      var targetOffsetHeight = rowsMeta.positions[elementIndex + 1] ? rowsMeta.positions[elementIndex + 1] - rowsMeta.positions[elementIndex] : rowsMeta.currentPageTotalHeight - rowsMeta.positions[elementIndex];
      scrollCoordinates.top = scrollIntoView({
        clientHeight: windowRef.current.clientHeight,
        scrollTop: windowRef.current.scrollTop,
        offsetHeight: targetOffsetHeight,
        offsetTop: rowsMeta.positions[elementIndex]
      });
    }

    scrollCoordinates = apiRef.current.unstable_applyPreProcessors('scrollToIndexes', scrollCoordinates, params);

    if (_typeof(scrollCoordinates.left) !== undefined || _typeof(scrollCoordinates.top) !== undefined) {
      apiRef.current.scroll(scrollCoordinates);
      return true;
    }

    return false;
  }, [logger, apiRef, windowRef, props.pagination]);
  var scroll = React.useCallback(function (params) {
    if (windowRef.current && params.left != null && colRef.current) {
      colRef.current.scrollLeft = params.left;
      windowRef.current.scrollLeft = params.left;
      logger.debug("Scrolling left: ".concat(params.left));
    }

    if (windowRef.current && params.top != null) {
      windowRef.current.scrollTop = params.top;
      logger.debug("Scrolling top: ".concat(params.top));
    }

    logger.debug("Scrolling, updating container, and viewport");
  }, [windowRef, colRef, logger]);
  var getScrollPosition = React.useCallback(function () {
    if (!(windowRef != null && windowRef.current)) {
      return {
        top: 0,
        left: 0
      };
    }

    return {
      top: windowRef.current.scrollTop,
      left: windowRef.current.scrollLeft
    };
  }, [windowRef]);
  var scrollApi = {
    scroll: scroll,
    scrollToIndexes: scrollToIndexes,
    getScrollPosition: getScrollPosition
  };
  useGridApiMethod(apiRef, scrollApi, 'GridScrollApi');
  var preventScroll = React.useCallback(function (event) {
    event.target.scrollLeft = 0;
    event.target.scrollTop = 0;
  }, []);
  useGridNativeEventListener(apiRef, function () {
    var _apiRef$current, _apiRef$current$rende, _apiRef$current$rende2;

    return (_apiRef$current = apiRef.current) == null ? void 0 : (_apiRef$current$rende = _apiRef$current.renderingZoneRef) == null ? void 0 : (_apiRef$current$rende2 = _apiRef$current$rende.current) == null ? void 0 : _apiRef$current$rende2.parentElement;
  }, 'scroll', preventScroll);
};
