// @mui/x-license-pro/node/utils/licenseInfo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LicenseInfo = void 0;

var _utils = require("@mui/utils");

/* eslint-disable no-underscore-dangle */
// Store the license information in a global so it can be shared
// when module duplication occurs. The duplication of the modules can happen
// if using multiple version of MUI X at the same time of the bundler
// decide to duplicate to improve the size of the chunks.
_utils.ponyfillGlobal.__MUI_LICENSE_INFO__ = _utils.ponyfillGlobal.__MUI_LICENSE_INFO__ || {
  key: undefined,
  releaseInfo: undefined
};

class LicenseInfo {
  static getKey() {
    return _utils.ponyfillGlobal.__MUI_LICENSE_INFO__.key;
  }

  static getReleaseInfo() {
    return _utils.ponyfillGlobal.__MUI_LICENSE_INFO__.releaseInfo;
  }

  static setLicenseKey(key) {
    _utils.ponyfillGlobal.__MUI_LICENSE_INFO__.key = key;
  }

  static setReleaseInfo(encodedReleaseInfo) {
    _utils.ponyfillGlobal.__MUI_LICENSE_INFO__.releaseInfo = encodedReleaseInfo;
  }

}

exports.LicenseInfo = LicenseInfo;
