// @mui/x-data-grid/node/constants/defaultGridSlotsComponents.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DATA_GRID_DEFAULT_SLOTS_COMPONENTS = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _Checkbox = _interopRequireDefault(require("@mui/material/Checkbox"));

var _TextField = _interopRequireDefault(require("@mui/material/TextField"));

var _FormControl = _interopRequireDefault(require("@mui/material/FormControl"));

var _Select = _interopRequireDefault(require("@mui/material/Select"));

var _Switch = _interopRequireDefault(require("@mui/material/Switch"));

var _Button = _interopRequireDefault(require("@mui/material/Button"));

var _Tooltip = _interopRequireDefault(require("@mui/material/Tooltip"));

var _Popper = _interopRequireDefault(require("@mui/material/Popper"));

var _components = require("../components");

var _GridColumnUnsortedIcon = require("../components/columnHeaders/GridColumnUnsortedIcon");

var _ErrorOverlay = require("../components/ErrorOverlay");

var _GridNoResultsOverlay = require("../components/GridNoResultsOverlay");

const DEFAULT_GRID_ICON_SLOTS_COMPONENTS = {
  BooleanCellTrueIcon: _components.GridCheckIcon,
  BooleanCellFalseIcon: _components.GridCloseIcon,
  ColumnMenuIcon: _components.GridTripleDotsVerticalIcon,
  OpenFilterButtonIcon: _components.GridFilterListIcon,
  FilterPanelDeleteIcon: _components.GridCloseIcon,
  ColumnFilteredIcon: _components.GridFilterAltIcon,
  ColumnSelectorIcon: _components.GridColumnIcon,
  ColumnUnsortedIcon: _GridColumnUnsortedIcon.GridColumnUnsortedIcon,
  ColumnSortedAscendingIcon: _components.GridArrowUpwardIcon,
  ColumnSortedDescendingIcon: _components.GridArrowDownwardIcon,
  ColumnResizeIcon: _components.GridSeparatorIcon,
  DensityCompactIcon: _components.GridViewHeadlineIcon,
  DensityStandardIcon: _components.GridTableRowsIcon,
  DensityComfortableIcon: _components.GridViewStreamIcon,
  ExportIcon: _components.GridSaveAltIcon,
  MoreActionsIcon: _components.GridMoreVertIcon,
  TreeDataCollapseIcon: _components.GridExpandMoreIcon,
  TreeDataExpandIcon: _components.GridKeyboardArrowRight,
  GroupingCriteriaCollapseIcon: _components.GridExpandMoreIcon,
  GroupingCriteriaExpandIcon: _components.GridKeyboardArrowRight,
  DetailPanelExpandIcon: _components.GridAddIcon,
  DetailPanelCollapseIcon: _components.GridRemoveIcon
};
/**
 * TODO: Differentiate community and pro value and interface
 */

const DATA_GRID_DEFAULT_SLOTS_COMPONENTS = (0, _extends2.default)({}, DEFAULT_GRID_ICON_SLOTS_COMPONENTS, {
  BaseCheckbox: _Checkbox.default,
  BaseTextField: _TextField.default,
  BaseFormControl: _FormControl.default,
  BaseSelect: _Select.default,
  BaseSwitch: _Switch.default,
  BaseButton: _Button.default,
  BaseTooltip: _Tooltip.default,
  BasePopper: _Popper.default,
  Cell: _components.GridCell,
  ColumnMenu: _components.GridColumnMenu,
  ErrorOverlay: _ErrorOverlay.ErrorOverlay,
  Footer: _components.GridFooter,
  Header: _components.GridHeader,
  Toolbar: null,
  PreferencesPanel: _components.GridPreferencesPanel,
  LoadingOverlay: _components.GridLoadingOverlay,
  NoResultsOverlay: _GridNoResultsOverlay.GridNoResultsOverlay,
  NoRowsOverlay: _components.GridNoRowsOverlay,
  Pagination: _components.GridPagination,
  FilterPanel: _components.GridFilterPanel,
  ColumnsPanel: _components.GridColumnsPanel,
  Panel: _components.GridPanel,
  Row: _components.GridRow
});
exports.DATA_GRID_DEFAULT_SLOTS_COMPONENTS = DATA_GRID_DEFAULT_SLOTS_COMPONENTS;
