// @docker/extension-preload/webpack.config.js
const path = require('path');

module.exports = {
  entry: path.resolve(__dirname, './src/index.ts'),
  mode: 'development',
  target: 'electron-preload',
  resolve: {
    extensions: ['.ts', '.js'],
  },
  module: {
    rules: [
      {
        // the basic styles defined in style.scss
        test: /\.scss$/,
        use: [
          { loader: 'style-loader', options: { injectType: 'lazyStyleTag' } },
          'css-loader',
          'sass-loader',
        ],
      },
      {
        // the font stuff from inside desktop-theme
        test: /\.css$/,
        use: [
          { loader: 'style-loader', options: { injectType: 'lazyStyleTag' } },
          'css-loader',
        ],
      },
      {
        // Inline so that we can encode everything inside the preload script.
        // A good next step would be to serve the font from a custom URL scheme
        // via Electron, and then just have the preload script reference the font
        // by URL, to limit the size of the preload script and stop its parsing being
        // such a bottleneck.
        test: /\.woff2$/,
        type: 'asset/inline',
      },
      {
        test: /\.ts$/,
        use: {
          loader: 'ts-loader',
          options: {
            projectReferences: true,
          },
        },
      },
    ],
  },
};
